### PR TITLE
Implemented `get_key_value`

### PR DESCRIPTION
**Pull request by jonathanGB**
Referring to #10

### CODE REVIEW

**Comment by jonhoo** [comment]
Excellent, thanks, let me take a look!

File: src/lib.rs [hunk 1]
```diff
@@ -403,6 +403,13 @@ where
         self.get(key, guard).map(then)
     }
 
+    /// Returns the key-value pair corresponding to the supplied key.
+    ///
+    /// Returns `None` if this map contains no mapping for `key`.
+    pub fn get_key_value<'g>(&'g self, key: &'g K, guard: &'g Guard) -> Option<(&'g K, &'g V)> {
```
Comment (jonhoo) [inline src/lib.rs]: I don't think you need the `'g` on the `self` reference here. As long as the guard lives, the pointers will remain valid, even if `self` completely goes away.

Comment (jonhoo) [reply src/lib.rs]: In fact, I don't think you need it on `key` either, since we return the key we read from the map.

Comment (jonhoo) [reply src/lib.rs]: Actually, damn, I take that back. I think there's a bug in `insert`. The `'g` _does_ have to be associated with the lifetime of `self` for [this](https://github.com/jonhoo/flurry/blob/6d56c58026f6edd0d9828ce48b3cbb6aecd38973/src/lib.rs#L1227) to be safe. It shouldn't be needed on `key` though!

File: src/lib.rs [hunk 2]
```diff
@@ -403,6 +403,13 @@ where
         self.get(key, guard).map(then)
     }
 
+    /// Returns the key-value pair corresponding to the supplied key.
+    ///
+    /// Returns `None` if this map contains no mapping for `key`.
```
Comment (jonhoo) [inline src/lib.rs]: It'd be good to match the documentation and bounds on [`HashMap::get_key_value`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get_key_value) here.

File: src/lib.rs (hunk 1 above)
Comment (jonhoo) [reply src/lib.rs]: Ah, I see, below you _are_ returning the `key` that is passed in. That is probably not what we want to do. Instead, we'll want an internal version of `get` that produces a `&Node`, and then have `get` call that and extract the key, and this have call that and extract both the key and the value.

Comment (cuviper) [inline src/lib.rs]: Following #25, this should probably use a borrowed `Q` type for the key.

Comment (jonathanGB) [reply src/lib.rs]: Yep, I had started changing everything to `&Q` but I see it's done :P

File: src/lib.rs [hunk 3]
```diff
@@ -432,7 +424,24 @@ where
         // next epoch after it is removed. since it wasn't removed, and the epoch was pinned, that
         // cannot be until after we drop our guard.
         let node = unsafe { node.deref() };
-        let node = node.as_node().unwrap();
+        node.as_node()
```
Comment (jonhoo) [inline src/lib.rs]: I think this should actually be
```rust
Some(node.as_node().expect("BinEntry::find should always return Node"))
```

Comment (jonathanGB) [reply src/lib.rs]: I had the callers of `get_node` use the `?` operator, but I'm fine with that as well

Comment (jonathanGB) [reply src/lib.rs]: Oh wait, no I kept it as an option because some sanity check in the method might return `None` (see lines 388 and 395).

Comment (jonhoo) [reply src/lib.rs]: Ah, no, this isn't about the return type of _this_ method (notice that we are re-wrapping in `Some`). It is about having an assertion that checks that `BinEntry::find` _indeed_ always returns a `Node`.

Comment (jonathanGB) [reply src/lib.rs]: (forget all I said)

**Pull request merged**